code/modules/projectiles/tgprojectile.dm

```diff
@@ -0,0 +1,1180 @@
+	var/homing_turn_speed = 10		//Angle per tick.
+	var/homing_inaccuracy_min = 0		//in pixels for these. offsets are set once when setting target.
+	var/homing_inaccuracy_max = 0
+	var/homing_offset_x = 0
+	var/homing_offset_y = 0
+
+	var/ignore_source_check = FALSE
+
+	var/damage = 10
+	var/damage_type = BRUTE //BRUTE, BURN, TOX, OXY, CLONE are the only things that should be in here
+	var/nodamage = 0 //Determines if the projectile will skip any damage inflictions
+	var/projectile_type = /obj/item/projectile
+	var/range = 50 //This will de-increment every step. When 0, it will deletze the projectile.
+	var/decayedRange			//stores original range
+	var/reflect_range_decrease = 5			//amount of original range that falls off when reflecting, so it doesn't go forever
+	var/reflectable = NONE // Can it be reflected or not?
+	
+	var/impact_effect_type //what type of impact effect to show when hitting something
+	var/log_override = FALSE //is this type spammed enough to not log? (KAs)
+
+	var/temporary_unstoppable_movement = FALSE
+
+/obj/item/projectile/Initialize()
+	. = ..()
+	permutated = list()
+	decayedRange = range
+
+/obj/item/projectile/proc/Range()
+	range--
+	if(range <= 0 && loc)
+		on_range()
+
+/obj/item/projectile/proc/on_range() //if we want there to be effects when they reach the end of their range
+	ammo.do_at_max_range(src)
+	qdel(src)
+
+//to get the correct limb (if any) for the projectile hit message
+/mob/living/proc/check_limb_hit(hit_zone)
+	return hit_zone
+
+/mob/living/carbon/check_limb_hit(hit_zone)
+	//message_admins("[hit_zone] zone")
+//TODO: make this actually do something
+	//if(get_bodypart(hit_zone))
+	//	return hit_zone
+	//else //when a limb is missing the damage is actually passed to the chest
+	return "chest"
+/obj/item/projectile/proc/prehit(atom/target)
+	return TRUE
+
+
+/obj/item/projectile/proc/vol_by_damage()
+	if(damage)
+		return CLAMP((damage) * 0.67, 30, 100)// Multiply projectile damage by 0.67, then CLAMP the value between 30 and 100
+	else
+		return 50 //if the projectile doesn't do damage, play its hitsound at 50% volume
+
+/obj/item/projectile/proc/on_ricochet(atom/A)
+	return
+
+/obj/item/projectile/proc/store_hitscan_collision(datum/point/pcache)
+	beam_segments[beam_index] = pcache
+	beam_index = pcache
+	beam_segments[beam_index] = null
+
+/atom/proc/handle_ricochet(obj/item/projectile/P)
+	return
+
+
+/obj/item/projectile/Bump(atom/A as mob|obj|turf|area)
+	var/datum/point/pcache = trajectory.copy_to()
+	var/turf/T = get_turf(A)
+	if(check_ricochet(A) && check_ricochet_flag(A) && ricochets < ricochets_max)
+		ricochets++
+		if(A.handle_ricochet(src))
+			on_ricochet(A)
+			ignore_source_check = TRUE
+			decayedRange = max(0, decayedRange - reflect_range_decrease)
+			range = decayedRange
+			if(hitscan)
+				store_hitscan_collision(pcache)
+			return TRUE
+
+	var/distance = get_dist(T, starting) // Get the distance between the turf shot from and the mob we hit and use that for the calculations.
```

Comment:
CM uses `distance_travelled`. We'll have two systems in parallel here. We need to determine which is more interesting/performant and keep only one.

Reply:
This is hitscan code, you know, for instant travel, like, laser beams.

Reply:
Either there is no hitscan cheeck or github is really messing up indentation here.
This is the universal projectile Bump code, which will trigger every time `Move()` meets an obstacle.

Reply:
Ah sorry, I misunderstood. the `distance` and `distance_travelled` are not the same. One is the actual distance, whereas the latter is the amount of tiles the projectile travelled. I don't know which one to use, I suppose I should stick with distance?

Reply:
`Distance` is from /tg/, `distance_travelled` is from CM.

Reply:
`distance_travelled` is used for accuracy checks, whereas `distance` is used for getting the `def_zone` which says what body part to aim at.

Reply:
Or at least that's what it should be doing. I'll just delete it as it currently does nothing.